Adds top level spider api with ioc, integrations, and configuration

electricjones opened the pull request:
This is a continuation of discussion #22. There is no rush here. #81 and #19 are probably the priority for now. Finally, I'm not married to any of these names or terms.

 _This PR is NOT ready to merge_. 

 This is a full and passing working draft, but could still use some cleanup, even if nothing changes structurally. Also note that I haven't updated the docs yet, but am going to detail everything here. I will update this comment as we come to decisions and it will turn into documentation before merging.

 Sorry, this is pretty long, but here we go.

---
### Goals for the Top-Level API
- [ ] Be able to configure spider in one location and then re-use that configuration
- [ ] Have sensible, cascading defaults for those configurations
- [ ] Pass that configuration to child objects, if using the Spider as a factory
- [ ] Allow user to handle `NotSupportedExceptions` #29 
- [ ] Include factories for query builders, connections, and more in the future
- [ ] Introduce a _light_ ioc container only into the top-level spider. Not required by components.
- [ ] Introduce an replaceable, optional event system. Also serves as an example for future integrations (caching, logging, etc).
  ### How Spider works
  - [ ] Configure a static via `Spider::setup()`
- [ ] Instantiate a new `$spider = Spider::make()` from default or other set connection
- [ ] That `$spider` extends the query builder `$spider->select->etc()`
- [ ] That `$spider` also acts as a factory:
- [ ] `$spider->connection()`, `$spider->querybuilder()` returns new, configured objects
- [ ] Anything used through `$spider` is passed optional configurations and the event dispatcher
- [ ] Those components use `ConfigurableTrait` and `UsesEventsTrait` which manage events and configuration. Note that eventing is optional. If the Connection Manager fires an event with no dispatcher set, that is fine
  ### Configuration
  
  Taking advice from @smolinari
- [ ] User MUST set a _default connection_ and at least one valid _connection_
- [ ] User MAY replace the default _ioc container_ with one that implements `Michaels\Manager\Contracts\IocContainerInterface` which only has a `fetch()` method
- [ ] User MAY replace the default _event dispatcher_ with any that implements `Spider\Integrations\Events\DispatcherInterface` (more on that below)
- [ ] User MAY define how _unsupported errors_ are handled (more on that below)
- [ ] User MAY use any _combination_ of the above as defaults will be provided
  
  ``` php
  Spider::setup([
     'connections' => [
         'default' => 'neo',
         'neo' => [
             'driver' => 'neo4j'
             // The rest of the parameters
         ],
     ],
     'integrations' => [ // optional
         'events' => $dispatcher // many ways to set dispatcher below
     ],
     'errors' => [
         'not_supported' => 'silent' // defaults to fatal, more below
     ]
  ], $optionalIocContainer);
  ```
  ### Instantiation and Factories
  
  Once the static instance is `setup`, the user can create instances
  
  ``` php
  $spider = Spider::make(); // default connection
  $spider = Spider::make('another_connection');
  ```
  
  Once an instance is made, you can also create components:
  
  ``` php
  $qb = $spider->querybuilder();
  $con = $spider->connection();
  // Or
  $qb = $spider->querybuilder('another_connection');
  $con = $spider->connection('another_connection');
  ```
  
  There are also a few managing methods
  
  ``` php
  $spider->getDriver()
  $spider->getConnection();
  $spider->getEventDispatcher();
  $spider->addConnection();
  ```
  - [ ] **Question 1**: Do we want a way to get the _same_ spider instance each time? `Spider::getInstance()`?
  ### Dependency Injection and Integrations
  
  This branch uses a modified version of my `michaels/data-manager` as a lightweight ioc container. I will release that version before this is merged.
  
  To avoid coupling the code too tightly, the IoC container never actually leaves the top level `$spider`. Instead, it only _holds_ objects and injects them into the constructors of the various components (Connections Manager, Query, Drivers, etc). This means that if someone is using the pieces independently of spider, they do not need to worry about the container, only set the dependencies _if they want to_.
  
  At the moment only events are used, but eventually I can picture caching, logging, and maybe profiling, among potential others.
  
  You can also set your own ioc container (say the one your framework uses), so long as it implements the interface.
  - [ ] **Question 2**: Would it be better to use duck-typing here? It may be difficult to wrap a frameworks container with a contract class. Is `fetch()` not standard enough?
  
  The supplied IoC container allows you to set dependencies in four ways:
  1. As a _string_ class name.
  2. As an already instantiated _object_
  3. As a closure _factory_
  4. As an instance of a _container_ with fetch. In this case, when the dep is needed, the library calls `$container->fetch('alias')`
  
  ``` php
  Spider::setup([
     'integrations' => [
         'events' => 'Some\Class\Name',
         'events' => $dispatcher, // or new Dispatcher()
         'events' => function () {
            return new Dispatcher();
         }
         'events' => $container, // with a registered dep called 'events'
     ]
  ]);
  ```
  ### Using Events
  
  At the moment, only the connection manager uses events. This was just done as a workflow example. Before this is merged (if we go forward with events now), we may need to move it into the rest of the system.
  
  To setup a component to use events, simply
  1. Include the trait
  2. Modify the constructor to accept and set a dispatcher
  3. Fire off events, confident that with no dispatcher, nothing will happen
  
  ``` php
  class SomeClass
  {
     use UsesEventsTrait;
  
     public function __construct(DispatcherInterface $events = null)
     {
         // Optional Event Dispatcher
         $this->setDispatcher($events);
     }
  
     public function someMethod()
     {
            $this->emit('my.event');
     }
  }
  ```
  
  You may also add events. This is shown in the tests.
  
  ``` php
  $events = [];
  $spider->getEventDispatcher()->addListener(
     'connections.manager.before_make',
     function () use (&$events) {
         $events[] = 'fired-before_make';
     });
  
  $spider->getEventDispatcher()->addListener(
     'connections.manager.after_make',
     function () use (&$events) {
         $events[] = 'fired-after_make';
     });
  
  $spider->connection(); // will fire both events
  
  $this->assertEquals('fired-before_make', $events[0], "failed to fire before");
  $this->assertEquals('fired-after_make', $events[1], 'failed to fire after');
  ```
  ## Configurable and NotSupported Traits
  
  This also introduces two new traits that add functionality to component parts.
  
  The `ConfigurableTrait` allows a component to hold optional configuration in a predictable way. It has an associated interface.
  - When instantiating the component, pass an optional `$configManager`, which can be an instance of `Michaels\Manager\Manager` or an array of configurations.
  - In the constructor, `setConfigMangaer()`. If not using config, then an empty manager is created
  - In the class, use `$this->config()->get('whatever')` to return the manager and get the item. See michaels/manager for all the options
  
  The `NotSupportedTrait` allows a component to declare a feature not supported and the trait handles how that issue is raise via exception or warning or silent. 
  
  The `NotSupportedTrait` requires the `ConfigurableTrait`.
  - `use Spider\Base\ConfigurableTrait`
  - `use Spider\Base\NotSupportedTrait`
  - When something is unsupported, simply `$this->notSupported($message)` and the trait will use the configuration to handle the error.
    - `'fatal'` throws a `NotSupportedException`
    - `'quiet'` raises an `E_USER_WARNING`
    - `'silent'` does nothing (this will probably lead to another failure down the chain.
    - defaults to fatal. We can change these names and add whatever we wish.
  
  Take a look at the `AbstractDriver` and one of the Driver implementations to see a full rundown.
  ### Open Questions and Tasks
  - [ ] **Task 1**: Split testing. Right now, all testing occurs in `Integrations/SpiderTest`, but should probably be split.
- [ ] **Task 2**: Better support of Event Testing, probably in its own test
- [ ] **Question 3**: Right now, it is possible to configure statically and use `make()` **or** to configure on an instance using `configure()` **or** to pass the configuration into the constructor. This seems flexible but obscure. We should probably settle on a 1 or two solid methods.
- [ ] **Question 4**: Is there a way to get rid of the `ThrowsNotSupportedStub` which is only used in the `ThrowsNotSupportedTest`?
- [ ] **Task 3**: When testing whether Spider passes not supported config and such, it relies on `getTree()` being unsupported, which it will not be forever.
- [ ] **Task 4**: Turn this PR into documentation and update docblocs.
- [ ] **Task 5**: Integrate Events, NotSupported, and Configuration throughout the system?

Of course, any feedback is welcome.

dmill-bz commented:
Just waned to let you know that I will eventually get around taking a look at this PR

electricjones commented:
That's fine. I figure this would be the last feature we merge into 0.4 before we bug scrub and cleanup for the release.

electricjones commented:
I simplified the way defaults are loaded and changed the `data-manager` dependency to the `develop` branch. I will soon release `0.8.5` and the dependency will change again.

Please `composer update`
